### ISSUE TITLE
Add encoding and compression to exports as well.

### ISSUE THREAD

**Issue by mkosler** [bjorn/tiled#343]
Specifically the Lua exporter. It would be nice to have those preferences reflect in the exports.

**Comment by tcoxon** [comment]
@bjorn: 
> I'm in a bit of a dilemma about whether those options should be separate, and if so, where to put them in the UI.

So this issue is blocked waiting on design work then?

I might be pressed to find a short-term solution for this problem (specifically for the JSON format) soon. If I was to create an interim solution and submit it as a pull request (although I can't promise I'll get permission/time to do this), is it likely to be accepted upstream? Or would you hold off on accepting pull requests until there's a proper design for how tile layer compression fits in with plugins?

What I'm thinking of is adding a "JSON with compression" file format option to the save/export dialogs. I _think_ that wouldn't need any changes outside of the JSON plugin.

**Comment by bjorn** [comment]
Actually the location in the UI is no longer an issue since the layer data format option has already moved from the preferences to the "New Map" dialog and the map properties view. The plugin can query it uses `Map::layerDataFormat` and write out different formats based on that. It should consider both `Map::XML` and `Map::CSV` as the current native array format (on reading I think this should be set to `Map::CSV` since I'd like to deprecate the `Map::XML` format.

So the only design related thing that would need to be solved is how to effectively share the code that packs and unpacks the tile layer data. This code is already inside of libtiled, but currently private to the `Tiled::MapWriter`. It's only about 20 lines though, so I'm fine with copying that around for now as well.

In fact, now that I have it all figured out I might as well implement it... unless you want to do it. :-)

**Comment by bjorn** [comment]
Removing this from Tiled 0.12 milestone since it's a bit too last-minute for my taste now given that implementing this would mean the default formats for Lua and JSON exports will change to zlib-compressed base64-encoded data, which many will be unprepared for. Therefor I'll do it after the release, then there's more time for communication and feedback.

**Comment by kamranayub** [comment]
Yah I was caught by this, didn't expect JSON to be compressed :(

**Comment by bjorn** [comment]
@kamranayub Sorry about that. At least it is easy to set the map to use the "CSV" format instead in the Properties view.
